diffkemp/simpll/InstPatternComparator.cpp

```diff
@@ -0,0 +1,660 @@
+    auto PatInst = PatBB->begin(), PatInstE = PatBB->end();
+    auto ModTerm = ModBB->getTerminator(), PatTerm = PatBB->getTerminator();
+
+    // Jump to the currently compared instruction pair.
+    jumpToInst(ModInst, ModPosition);
+    jumpToInst(PatInst, PatPosition);
+
+    while (ModInst != ModInstE && PatInst != PatInstE) {
+        ModPosition = &*ModInst;
+        PatPosition = &*PatInst;
+
+        // Check whether the compared pattern ends at this instruction.
+        if (hasPatternEnd(&*PatInst))
+            return 0;
+
+        // If at the end of only one basic block, leave the rest to
+        // unconditionally connected basic blocks (if there are any).
+        if (&*ModInst == ModTerm && &*PatInst != PatTerm
+            && ModTerm->getNumSuccessors() == 1) {
+            return 1;
+        }
+        if (&*ModInst != ModTerm && &*PatInst == PatTerm) {
+            if (PatTerm->getNumSuccessors() == 1)
+                return 0;
+
+            // If no unconditional successor exists, skip to the terminator
+            // instruction.
+            ModInst = ModInstE;
+            --ModInst;
+            continue;
+        }
+
+        // Compare current instructions with operands.
+        if (int Res = cmpOperationsWithOperands(&*ModInst, &*PatInst)) {
+            // Remove newly added value and input mappings.
+            eraseNewlyMapped();
+
+            // When in an instruction group, do not allow module instruction
+            // skipping.
+            if (GroupDepth > 0)
+                return Res;
+
+            // Skip the module instruction.
+            ++ModInst;
+            continue;
+        }
+
+        // Update the depth of pattern instruction groups.
+        updateGroupDepth(&*PatInst);
+
+        ++ModInst;
+        ++PatInst;
+    }
+
+    if (ModInst != ModInstE && PatInst == PatInstE)
+        return 1;
+    if (ModInst == ModInstE && PatInst != PatInstE)
+        return -1;
+
+    return 0;
+}
+
+/// Compare global values by their names, because their indexes are not
+/// expected to be the same.
+int InstPatternComparator::cmpGlobalValues(GlobalValue *ModVal,
+                                           GlobalValue *PatVal) const {
+    if (ModVal->hasName() && PatVal->hasName()) {
+        // Both values are named, compare them by names
+        auto ModName = ModVal->getName();
+        auto PatName = PatVal->getName();
+
+        if (namesMatch(ModName, PatName, IsLeftSide))
+            return 0;
+
+        return 1;
+    }
+
+    return ModVal != PatVal;
+}
+
+/// Compare a module value with a pattern value using serial numbers.
+/// Note: Parts of this function have been adapted from FunctionComparator.
+/// Therefore, LLVM licensing also applies here. See the LICENSE information
+/// in the appropriate llvm-lib subdirectory for more details.
+int InstPatternComparator::cmpValues(const Value *ModVal,
+                                     const Value *PatVal) const {
```

Comment:
Should be possible but will probably require a search operation.

Reply:
This would require us to perform a search on the synchronisation maps. Not sure that is possible since we do not know the serial numbers.

Reply:
So this wouldn't work?
```cpp
int InstPatternComparator::cmpValues(... ModVal, ... PatVal) {
    if (int Res = FunctionComparator::cmpValues(ModVal, PatVal))
        return Res;

    if (sn_mapL[ModVal] == sn_mapL.size() - 1)
        NewlyMappedModValues.insert(ModVal);
    // ... same for PatVal

     if (int Res = mapInputValues(ModVal, PatVal))
        return Res;

    return 0;
}
```

Reply:
Right. For some reason I thought the maps use `Value* : int` pairs as keys which simple does not make sense (and would require us to know the assigned numbers before each lookup). Since that is not the case, this should easily be doable.